## in this project we will work with breast cancer dataset from sklearn
## our porblem is to classify the cancer as malignant or benign

In [108]:
import numpy as np
import pandas as pd
from sklearn.model_selection import train_test_split
import matplotlib.pyplot as plt
from sklearn.datasets import load_breast_cancer
from sklearn.preprocessing import StandardScaler
from sklearn.metrics import accuracy_score,classification_report, confusion_matrix, recall_score, precision_score
from sklearn.linear_model import LogisticRegression
from sklearn.svm import LinearSVC
from sklearn.tree import DecisionTreeClassifier
from sklearn.ensemble import RandomForestClassifier
from sklearn.naive_bayes import GaussianNB
from sklearn.neighbors import KNeighborsClassifier



np.random.seed(42)

In [109]:
df = load_breast_cancer(as_frame=True)

In [110]:
df['data'].columns

Index(['mean radius', 'mean texture', 'mean perimeter', 'mean area',
       'mean smoothness', 'mean compactness', 'mean concavity',
       'mean concave points', 'mean symmetry', 'mean fractal dimension',
       'radius error', 'texture error', 'perimeter error', 'area error',
       'smoothness error', 'compactness error', 'concavity error',
       'concave points error', 'symmetry error', 'fractal dimension error',
       'worst radius', 'worst texture', 'worst perimeter', 'worst area',
       'worst smoothness', 'worst compactness', 'worst concavity',
       'worst concave points', 'worst symmetry', 'worst fractal dimension'],
      dtype='object')

In [111]:
df['data'].info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 569 entries, 0 to 568
Data columns (total 30 columns):
 #   Column                   Non-Null Count  Dtype  
---  ------                   --------------  -----  
 0   mean radius              569 non-null    float64
 1   mean texture             569 non-null    float64
 2   mean perimeter           569 non-null    float64
 3   mean area                569 non-null    float64
 4   mean smoothness          569 non-null    float64
 5   mean compactness         569 non-null    float64
 6   mean concavity           569 non-null    float64
 7   mean concave points      569 non-null    float64
 8   mean symmetry            569 non-null    float64
 9   mean fractal dimension   569 non-null    float64
 10  radius error             569 non-null    float64
 11  texture error            569 non-null    float64
 12  perimeter error          569 non-null    float64
 13  area error               569 non-null    float64
 14  smoothness error         5

In [112]:
df['data'].head()

mean radius  mean texture  mean perimeter  mean area  mean smoothness  \
0        17.99         10.38          122.80     1001.0          0.11840   
1        20.57         17.77          132.90     1326.0          0.08474   
2        19.69         21.25          130.00     1203.0          0.10960   
3        11.42         20.38           77.58      386.1          0.14250   
4        20.29         14.34          135.10     1297.0          0.10030   

   mean compactness  mean concavity  mean concave points  mean symmetry  \
0           0.27760          0.3001              0.14710         0.2419   
1           0.07864          0.0869              0.07017         0.1812   
2           0.15990          0.1974              0.12790         0.2069   
3           0.28390          0.2414              0.10520         0.2597   
4           0.13280          0.1980              0.10430         0.1809   

   mean fractal dimension  ...  worst radius  worst texture  worst perimeter  \
0                 0.07871  ...         25.38          17.33           184.60   
1                 0.05667  ...         24.99          23.41           158.80   
2                 0.05999  ...         23.57          25.53           152.50   
3                 0.09744  ...         14.91          26.50            98.87   
4                 0.05883  ...         22.54          16.67           152.20   

   worst area  worst smoothness  worst compactness  worst concavity  \
0      2019.0            0.1622             0.6656           0.7119   
1      1956.0            0.1238             0.1866           0.2416   
2      1709.0            0.1444             0.4245           0.4504   
3       567.7            0.2098             0.8663           0.6869   
4      1575.0            0.1374             0.2050           0.4000   

   worst concave points  worst symmetry  worst fractal dimension  
0                0.2654          0.4601                  0.11890  
1                0.1860          0.2750                  0.08902  
2                0.2430          0.3613                  0.08758  
3                0.2575          0.6638                  0.17300  
4                0.1625          0.2364                  0.07678  

[5 rows x 30 columns]

In [113]:
df['target'].head()

0    0
1    0
2    0
3    0
4    0
Name: target, dtype: int64

In [114]:
df['target'].info()

<class 'pandas.core.series.Series'>
RangeIndex: 569 entries, 0 to 568
Series name: target
Non-Null Count  Dtype
--------------  -----
569 non-null    int64
dtypes: int64(1)
memory usage: 4.6 KB


In [115]:
df['target'].value_counts()

target
1    357
0    212
Name: count, dtype: int64

## so we have 357 malignant tumors labeled as 1, and 257 benign as 0

## split the target and features

In [116]:
X = df['data']
y = df['target']

## split the train and test set.
## set the test size to 25% and 75% for the train

In [117]:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.25, random_state=42)

In [118]:
X_train.sample(), X_test.sample()

(     mean radius  mean texture  mean perimeter  mean area  mean smoothness  \
 466        13.14         20.74           85.98      536.9          0.08675   
 
      mean compactness  mean concavity  mean concave points  mean symmetry  \
 466            0.1089          0.1085               0.0351         0.1562   
 
      mean fractal dimension  ...  worst radius  worst texture  \
 466                  0.0602  ...          14.8          25.46   
 
      worst perimeter  worst area  worst smoothness  worst compactness  \
 466            100.9       689.1            0.1351             0.3549   
 
      worst concavity  worst concave points  worst symmetry  \
 466           0.4504                0.1181          0.2563   
 
      worst fractal dimension  
 466                  0.08174  
 
 [1 rows x 30 columns],
      mean radius  mean texture  mean perimeter  mean area  mean smoothness  \
 249        11.52         14.93           73.87      406.3           0.1013   
 
      mean compactne

## Normalize the data for numerical stability

In [119]:
train_scaler = StandardScaler()
X_train = train_scaler.fit_transform(X_train)

test_scaler = StandardScaler()
X_test = test_scaler.fit_transform(X_test)

## fit a logistic regression model to the training set

In [120]:
log_model = LogisticRegression()

log_model.fit(X_train, y_train)

pred = log_model.predict(X_test)

## Evalute the model

In [121]:
TN, FP, FN, TP = confusion_matrix(y_test, pred).ravel()

print("True Positive (TP): ", TP)
print("False Positive (FP): ", FP)
print("True Negative(TN): ", TN)
print("False Negatie (FN): ", FN)

True Positive (TP):  89
False Positive (FP):  2
True Negative(TN):  52
False Negatie (FN):  0


In [122]:
conf_m = confusion_matrix(y_test, pred)
accuracy = accuracy_score(y_test, pred)
c_report = classification_report(y_test, pred)

In [123]:
print(f"Accuracy: {accuracy * 100:.2f}%")
print("Classification Report:\n", c_report)
print("Confusion Matrix:\n", conf_m)

Accuracy: 98.60%
Classification Report:
               precision    recall  f1-score   support

           0       1.00      0.96      0.98        54
           1       0.98      1.00      0.99        89

    accuracy                           0.99       143
   macro avg       0.99      0.98      0.99       143
weighted avg       0.99      0.99      0.99       143

Confusion Matrix:
 [[52  2]
 [ 0 89]]


## we can also calculate the accuracy this way.

In [124]:
accuracy_2 = (TP + TN) / (TP + TN + FP + FN )
print(f"Accuracy: {accuracy_2 * 100:.2f}%")


Accuracy: 98.60%


## now we will use a bunch of binary classifiers in the scikit-learn library.

In [125]:
# we will store the models in one array to quickly loop over them
models = {}

models['Logistic Regression'] = LogisticRegression()

models['Linear SVC'] = LinearSVC()

models['Decision Tree'] = DecisionTreeClassifier()

models['Random Forest'] = RandomForestClassifier()

models['Gassian NB'] = GaussianNB()

models['K Neighbors'] = KNeighborsClassifier()

In [126]:
accuracy, precision, recall = {}, {}, {}

for key in models.keys():

    # fit x & y train
    models[key].fit(X_train, y_train)

    # predict
    predictions = models[key].predict(X_test)

    # scores
    accuracy[key] = accuracy_score(predictions, y_test)
    precision[key] = precision_score(predictions, y_test)
    recall[key] = recall_score(predictions, y_test)



In [127]:
df_model = pd.DataFrame(index=models.keys(), columns=['Accuracy','Precision','Recall'])
df_model['Accuracy'] = accuracy.values()
df_model['Precision'] = precision.values()
df_model['Recall'] = recall.values()

In [129]:
print(df_model)

                     Accuracy  Precision    Recall
Logistic Regression  0.986014   1.000000  0.978022
Linear SVC           0.972028   0.988764  0.967033
Decision Tree        0.944056   0.977528  0.935484
Random Forest        0.972028   0.988764  0.967033
Gassian NB           0.951049   0.977528  0.945652
K Neighbors          0.958042   0.988764  0.946237
